Use MPI_Init_thread if OpenMP is enabled

cniethammer opened the pull request:
If MPI is used in a multi-threaded environment the MPI implementation has to be aware of the calling context of MPI methods. Therefore, initialize the MPI environment with MPI_Init_thread requesting the desired thread support level. So far ls1 makes MPI calls only from the main thread, i.e., uses MPI_THREAD_FUNNELED. For details see the MPI standard.

cniethammer committed:
Use MPI_Init_thread requesting the necessary thread support level

If MPI is used in a multi-threaded environment the MPI implementation has
to be aware of the calling context of MPI methods. Therefore, initialize
the MPI environment with MPI_Init_thread requesting the desired thread
support level. So far, ls1 built with OpenMP support makes MPI calls only
from the main thread, i.e., uses MPI_THREAD_FUNNELED. For details see the
MPI standard.

Signed-off-by: Christoph Niethammer <niethammer@hlrs.de>

FG-TUM commented on src/MarDyn.cpp:
```diff
@@ -132,8 +132,18 @@ int run_unit_tests(const Values &options, const std::vector<std::string> &args)
 int main(int argc, char** argv) {
 #ifdef ENABLE_MPI
+#if defined(_OPENMP)
+	int thread_support_level_requested = MPI_THREAD_FUNNELED; /* So far only the main thread makes MPI calls */
```
```suggestion
	const int thread_support_level_requested = MPI_THREAD_FUNNELED; /* So far only the main thread makes MPI calls */
```

FG-TUM commented on src/MarDyn.cpp:
```diff
@@ -132,8 +132,18 @@ int run_unit_tests(const Values &options, const std::vector<std::string> &args)
 #ifdef ENABLE_MPI
+#if defined(_OPENMP)
+	int thread_support_level_requested = MPI_THREAD_FUNNELED; /* So far only the main thread makes MPI calls */
+	int thread_support_level_provided = MPI_THREAD_SINGLE;
```
I don't quite get why this is set to `MPI_THREAD_SINGLE` if the next line immediately overwrites this value. As far as I understand, `MPI_Init_thread` will always set the last parameter to something, hence, I'd rather suggest default initialization to not give of the impression that the initial value is actually used.

```suggestion
	int thread_support_level_provided{};
```

cniethammer replied:
`MPI_THREAD_SINGLE` means "no thread support", therefore make this the initialization to be safe in any case. Note: As long as there is no ABI for MPI (in preparation) you have no guarantees about the integer values for the MPI thread support level constants, i.e., zero could be something unexpected.

FG-TUM replied:
What do you mean by "in any case"? According to the documentation `MPI_Init_thread` is guaranteed to overwrite this value, hence it shouldn't matter how we initialize it, hence I would argue we shouldn't initialize it to something misleading.
But if you want to keep the value at least add a comment that the value will be overwritten immediately.

FG-TUM commented on src/MarDyn.cpp:
```diff
@@ -132,8 +132,18 @@ int run_unit_tests(const Values &options, const std::vector<std::string> &args)
+	MPI_Init_thread(&argc, &argv, thread_support_level_requested, &thread_support_level_provided);
+	if (thread_support_level_provided < thread_support_level_requested) {
+		std::cerr << "Error: MPI implementation does not support the requested thread support level." << std::endl;
+		std::exit(EXIT_FAILURE);
```
Shouldn't we still call some `MPI` cleanup function? How about:
```suggestion
		mardyn_exit(EXIT_FAILURE);
```

FG-TUM commented on src/MarDyn.cpp:
```diff
@@ -132,8 +132,18 @@ int run_unit_tests(const Values &options, const std::vector<std::string> &args)
  */
 int main(int argc, char** argv) {
 #ifdef ENABLE_MPI
+#if defined(_OPENMP)
```
any reason not to merge the two?
```suggestion
#if defined(ENABLE_MPI) && defined(_OPENMP)
```

FG-TUM replied:
Completely correct. Disregard this one.

FG-TUM reviewed (commented):
Just a bit of readability

cniethammer commented:
Based on the feedback and diving deeper into the MPI standard, I refactored this PR.

Note:
* The logging infrastructure is not available at this point in the code. Therefore `mardyn_exit()` cannot be used, because it uses the Logger internally.
* I replaced `MPI_Init()` with `MPI_Init_thread()` entirely now. The MPI standard says here 'A call to MPI_INIT has the same effect as a call to MPI_INIT_THREAD with a required = MPI_THREAD_SINGLE." (MPI 4.1 p.486)

HomesGH commented on src/MarDyn.cpp:
```diff
@@ -132,8 +132,20 @@ int run_unit_tests(const Values &options, const std::vector<std::string> &args)
-	MPI_Init(&argc, &argv);
-#endif
+#if defined(_OPENMP)
+	const int thread_support_level_requested = MPI_THREAD_FUNNELED; /* So far only the main thread makes MPI calls */
```
Out of interest:
How do you know that in ls1 on the main thread makes MPI calls? Because nowhere in the code there is `#pragma omp parallel` and MPI combined?

cniethammer replied:
I went through the codebase looking for all places with `#pragma omp parallel`...
Nearly all MPI stuff is encapsulated and outside the loops that use OpenMP. The only place I found where MPI might be affected by OpenMP was in the `NonBlockingMPIMultiStepHandler` if `ADVANCED_OVERLAPPING` is enabled. However, here it is protected with a respective `omp master` ensuring that only the primary thread runs this part.

Also, if there was any problematic place, the code would have had issues with the existing `MPI_Init`. ;)

cniethammer commented:
@HomesGH Can we merge this after the taking into account our comment and changes?

HomesGH commented:
> @HomesGH Can we merge this after the taking into account our comment and changes?

I am fine with this PR but note [this open conversation](https://github.com/ls1mardyn/ls1-mardyn/pull/306#discussion_r1613543334) by @FG-TUM .

FG-TUM commented on src/MarDyn.cpp:
```diff
@@ -132,8 +132,20 @@ int run_unit_tests(const Values &options, const std::vector<std::string> &args)
+	MPI_Init_thread(&argc, &argv, thread_support_level_requested, &thread_support_level_provided);
+	if (thread_support_level_provided < thread_support_level_requested) {
+		/* The logging infrastructure is not initalized yet, therfore perform manual output and exit. */
+		std::cerr << "ERROR: ls1 was build with OpenMP. However, the MPI implementation does not provide the necessary thread support level." << std::endl;
+		MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
+		::exit(EXIT_FAILURE);
```
```suggestion
		MARDYN_EXIT("ERROR: ls1 was build with OpenMP. However, the MPI implementation does not provide the necessary thread support level.");
```

Please use the exit macro for consistent exit behavior.

cniethammer replied:
We cannot use the `mardyn_exit` macro as it relies on the logger infrastructure, which is not initialized at this point in the program (and may also dependson the MPI part). However, IMHO, we are so early in the program, that we do not need the extensive logger.

FG-TUM replied:
> However, IMHO, we are so early in the program, that we do not need the extensive logger.

I respectfully disagree, especially when looking at your related PR #358. Adding exceptions to a pattern like this makes the code more confusing, prone to set bad precedent for new developers, and needs clear documentation.
I'll add a suggestion on how to tackle the issue in the mentioned PR.

cniethammer replied:
OK, then I'm looking forward to your suggestion to address the Logger issues in a cleaner way

cniethammer replied:
Thanks for the inspiration!
I will update this PR when https://github.com/ls1mardyn/ls1-mardyn/pull/359 is merged.

cniethammer replied:
Did not find time yet but will see to do so in the next weeks.

FG-TUM commented on src/MarDyn.cpp:
```diff
@@ -132,8 +132,20 @@ int run_unit_tests(const Values &options, const std::vector<std::string> &args)
 #ifdef ENABLE_MPI
-	MPI_Init(&argc, &argv);
-#endif
+#if defined(_OPENMP)
+	const int thread_support_level_requested = MPI_THREAD_FUNNELED; /* So far only the main thread makes MPI calls */
+#else
+	const int thread_support_level_requested = MPI_THREAD_SINGLE; /* No thread support required */
+#endif /* _OPENMP */
```
```suggestion

	const int thread_support_level_requested = 
#if defined(_OPENMP)
		MPI_THREAD_FUNNELED; /* So far only the main thread makes MPI calls */
#else
		MPI_THREAD_SINGLE; /* No thread support required */
#endif /* _OPENMP */
```

Arguably a question of style, but I prefer to `ifdef` only what is really different.

cniethammer replied:
I never thought of this way to write it - but I must say I find it much harder to read as the variable name is not in the same line as the value, which makes my eyes jump over more lines of code than I want and my brain has to keep track of the ifdef in the meanwhile ...

FG-TUM reviewed (changes requested):
One style suggestion and one important change.
